### PR TITLE
View CANs

### PR DESCRIPTION
**Pull request by carjug**
## Issue
#93 

## What changed
- Add model for CANInfo
- Add model for CANAmount
- Add Django Admin view for combined CANs data (see screenshot)
- Adds migrations for the two models
- Adds fixture data for CANInfo and CANAmount models

## How to review (for devs)
Compare schema in Django migration against Google Sheet model
Compare fake data in fixture to Airtable prototype
Run project locally and review Persons page in Django Admin

## Screenshots
Screen Shot 2021-09-17 at 6 20 03 PM
Screen Shot 2021-09-17 at 6 20 19 PM
Screen Shot 2021-09-17 at 6 20 38 PM

### CODE REVIEW

File: opre_ops/ops_site/models.py
Comment (amymok) [inline opre_ops/ops_site/models.py:42]: (nitpick) a newline here.  Although we haven't set up circleci, I wonder if we want to run flake8 before we merge in a pull request manually.  I don't think there's any bad styling, just that it may snowball into a bigger task if we haven't done it since the beginning.

**Review by amymok — APPROVED** [review]
LGTM!

File: opre_ops/ops_site/models.py
Comment (alexsoble) [inline opre_ops/ops_site/models.py:42]: [_suggestion_]: A docstring comment here explaining what a CAN is, plus any other important information that a dev working on this model should know about the CANInfo model.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [inline opre_ops/ops_site/migrations/0007_canamount_caninfo.py:23]: [_praise_] The Google Sheet describing this table says "On_delete(SET_NULL) or (PROTECT)". I think Protect was the right choice here! In general I see us using soft delete strategies, where we retain records in the database and store some state like "active"/"inactive" rather than hard-deleting database entries. I like the idea of an error being raised if an operation tries to delete an agency referenced by CAN data.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [inline opre_ops/ops_site/migrations/0007_canamount_caninfo.py:34]: [_question_] The Google Sheet spec says DateField, not CharField. Which do we want to use here, and why? 

Are there advantages to CharField? If we go with CharField, should the max_length be 4 and not 5? 

My assumption is that using a DateField (or even an integer) would let us do math — e.g. `+= 1 year` operations — which we can't do on CharFields.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [inline opre_ops/ops_site/migrations/0007_canamount_caninfo.py:37]: [_thought, non-blocking_]: I notice that the Google Sheet spec doesn't specify a maximum amount for these values, but this implementation does. This would restrict these database fields to amounts below $100 million, because the decimal places count towards the maximum. 

$100 million or below seems reasonable. If I were the implementer I might have set the maximum higher? Non-blocking, easy enough to adjust if we find that the real data includes larger sums than we expected.

File: opre_ops/ops_site/models.py
Comment (alexsoble) [inline opre_ops/ops_site/models.py:55]: [_suggestion_]: Parallel suggestion to the one above ~ a docstring comment explaining what a CAN is, plus any other important information that a dev working on this model should know about the CANAmount model.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [inline opre_ops/ops_site/migrations/0007_canamount_caninfo.py:21]: [_issue, blocking_]: Following up on our conversation in Slack, suggest that we use `blank=True` here instead. This should resolve the issue we saw where CANs fail to save in Django Admin.

https://docs.djangoproject.com/en/3.2/ref/models/fields/#null
https://docs.djangoproject.com/en/3.2/ref/models/fields/#blank

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [inline opre_ops/ops_site/migrations/0007_canamount_caninfo.py:38]: [_issue, blocking_]: Similar to the issue about the `purpose` field above. Suggest that we use `blank=True` here. This should resolve an issue where CAN amounts fail to save in Django Admin.

https://docs.djangoproject.com/en/3.2/ref/models/fields/#null
https://docs.djangoproject.com/en/3.2/ref/models/fields/#blank

**Review by alexsoble — CHANGES_REQUESTED** [review]

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py:34]: Ah yes, I meant to ask the group about this. I'm not positive what we should do here. We'll never use a full date for just the year. If we need to do math on the year, I can possibly be convinced, but the DateField requires a month and day to be specified, which to me did not seem like something we would want or need. But, yes, open to discussion!

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py:37]: Ah, wait, I see. The decimal places gets factored into the max digits (didn't realize this).

Comment (alexsoble) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py:37]: @carjug 👍 👍 on `max_digits=12`

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py:34]: Would be interested to hear @amymok's thoughts when she is back in the office because she authored the Google Sheet spec. 

I wonder if integer gives us the best of all worlds — you can get the next fiscal year given a current year if you need it (`next = current + 1`), no extraneous month or day data involved, some built-in validation:

```python
>>> int('garbage input')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'garbage input'
```

Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py:34]: @amymok do you have thoughts here?

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: @alexsoble I think this is resolved now, can you double check

**Comment by carjug** [comment]
Screen Shot 2021-09-22 at 6 53 10 PM
Screen Shot 2021-09-22 at 6 53 18 PM

@amymok @alexsoble _I think_ I addressed most of the review feedback, plus the requested changes from Nina here. Can you all run through the changes again and LMK? Still wondering on Datefield question for `fiscal_year` but everything else should be resolved, I believe.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: Yep, this looks resolved when I try editing/saving CANs locally!

File: Pipfile
Comment (alexsoble) [inline Pipfile:9]: [_suggestions, blocking_]: I notice a new library introduced in the last round of changes. Two notes/suggestions here:

1. Importing the library with an asterisk in place a version means that no version is being pinned. This means that updating our dependencies can introduce breaking changes without us knowing. Let's pin to the latest major version when we introduce a new library. The Eng Practices Guide has a little bit about pinning dependencies here: https://engineering.18f.gov/python/.
2. It looks like the only import from this library is `Choices`. The [django-model-utils docs for Choices](https://django-model-utils.readthedocs.io/en/latest/utilities.html#choices) say: "Django 3.0 adds enumeration types. These provide most of the same features as Choices." Our project is using Django 3.2.6, so I would suggest we use Django-provided built-in `Choices` rather than a module from an external library. If there are specific features we need from django-model-utils that we can't get straight from Django, let's document those. Otherwise, the fewer external libraries we use the better off we are in terms of both security and complexity. 

Summary: I would suggest we remove the library if possible and use built-in Django features instead. If the library is needed, let's pin to a major version.

**Review by alexsoble — CHANGES_REQUESTED** [review]

**Comment by alexsoble** [comment]
@carjug All of my review feedback from the first round is addressed except for the discussion about the fiscal year data type.  I also left suggestions about the introduction of a new library since my first review. So, nearly there! Just two things to resolve from my point of view.

File: Pipfile
Comment (carjug) [reply Pipfile:9]: Ok sounds good. I didn't know about the enumeration types in Django 3.0. Will make those changes now.

**Comment by carjug** [comment]
@alexsoble yep just waiting to hear from @amymok about the `fiscal_year`.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (amymok) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: Oh oops, definitely missed this.  I chose DateField under the impression that it gives us more validation out of the box.  But since we are only using the year field, it may not be necessary because we probably don't need time delta or any of those.  So we can be flexible.  If we are giving the choices of CharField and IntegerField.  IntegerField would be better both for us to increment as well as to validate.

Here're some resources on how people have done it with choices:

- https://newbedev.com/year-field-in-django
- https://stackoverflow.com/questions/1517474/only-showing-year-in-django-admin-a-yearfield-instead-of-datefield/24656072#24656072

File: Pipfile
Comment (carjug) [reply Pipfile]: @alexsoble so I just added the list of choices tuples back in that was there in the original code for the PR and moved them inside the relevant class. I was using `Choices` from the imported library to try and simplify the code (AKA not need a tuple for a single choice, and instead just a single string) but I understand your concern about unnecessary libraries. If I were to use an Enumeration class such as TextChoices, IMO the benefits do not outweigh the overly complicatedness and, also, the TextChoices class gives me less control over how things are displayed in the UI, i.e. Django automatically uppercases the first letter and lower cases the rest so what should display as `OPRE Appropriation` displays as `Opre Appropriation` in the UI and it's more complicated to overwrite this than it's worth. So back to the simple way!

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: Ok done! Changed to integer field/

Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: I actually ended up changing it to a DecimalField so that I could set a max_digits on it, max_length does not work on IntegerField.

File: opre_ops/ops_site/admin.py
Comment (amymok) [inline opre_ops/ops_site/admin.py:53]: [thought][non-blocking] There's something in Python 3 that called [f-string](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/).  It makes the string looks cleaner.

File: opre_ops/ops_site/migrations/0007_auto_20210923_1822.py
Comment (amymok) [inline opre_ops/ops_site/migrations/0007_auto_20210923_1822.py:39]: [suggestion][blocking]: The [resource here](https://newbedev.com/year-field-in-django) suggests the use of validating and bounding the year.  We should consider using some form of that, so we can prevent numbers that are not year to be rejected before saving the record.

File: opre_ops/ops_site/models.py
Comment (amymok) [inline opre_ops/ops_site/models.py:76]: [suggestion][blocking]: The resource here suggests the use of validating and bounding the year. We should consider using some form of that, so we can prevent numbers that are not year to be rejected before saving the record.

Comment (carjug) [reply opre_ops/ops_site/models.py:76]: @amymok sounds good, what year do you think we should set as the upper and lower limits??

File: Pipfile
Comment (alexsoble) [reply Pipfile]: @carjug Gotcha! I haven't done a deep dive and might be wrong, but it looks like [TextChoices](https://docs.djangoproject.com/en/3.0/ref/models/fields/#enumeration-types) could give you more control over how choices are displayed, by surfacing a human readable `label` option? 

I'll marked this as resolved for now, since I am happy with any solution that uses either no new libraries, or a new library pinned to a specific version. 

Thank you for implementing this feedback @carjug!

Comment (carjug) [reply Pipfile]: Yeah I tried using the `label` option, but it was not working as one would expect.

File: opre_ops/ops_site/models.py
Comment (alexsoble) [reply opre_ops/ops_site/models.py:76]: Already posted this in Slack, but I'm not sure this suggestion should be a blocking one. The data model we are implementing here doesn't say anything about a max and a min value for fiscal years. Seems to me that data validation for fiscal years would fall into a different ticket / scope of work.

File: opre_ops/ops_site/admin.py
Comment (alexsoble) [reply opre_ops/ops_site/admin.py:53]: Fwiw I like this non-blocking suggestion, f-strings make interpolated strings like this much easier to read.

File: opre_ops/ops_site/models.py
Comment (amymok) [reply opre_ops/ops_site/models.py:76]: I am ok with this as well.

File: opre_ops/ops_site/migrations/0007_canamount_caninfo.py
Comment (alexsoble) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: My first initial reaction was that DecimalField with `decimal_places=0` seems a little less idiomatic than an IntegerField for storing years. I pulled down this latest code using DecimalField and noticed that Django Admin formats the fiscal year correctly. No weird decimal places anywhere. I was able to increment the fiscal year seamlessly using the Django Admin UI. And I see how it gets us some easy-to-implement `max_length` validation @carjug. So I don't have any blocking suggestions or thoughts here! This isn't exactly what's in the Google Sheets spec but it doesn't seem unreasonable. Open to your input/feedback as one of the co-designers of the data model @amymok.

Comment (carjug) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: The reason I used DecimaField instead of Integer is that IntegerField does not have a way to set a max_length. You can pass in `max_length` as a kwarg to `IntegerField` but it gets ignored. @alexsoble

Comment (amymok) [reply opre_ops/ops_site/migrations/0007_canamount_caninfo.py]: I would suggest we use `IntegerField` for a few reasons, (but also do not have the exact proofs though):
- it is more idiomatic to use `IntegerField`
- there may be unnecessary performance hit when we do comparison and filtering when using `DecimalField`
- we can use the `MidValueValidator` and `MaxValueValidator` to bound the field once we figure out what validation we want to do, I would not worry about needing it to have a `max_length` right now because there are still a big gap on what really is the validate numbers within that

File: opre_ops/ops_site/admin.py
Comment (amymok) [inline opre_ops/ops_site/admin.py:53]: [suggestion][blocking] removing `str()` should work

Comment (carjug) [reply opre_ops/ops_site/admin.py]: Ok should be good @amymok . If it looks good can you 👍 approve :)

**Comment by carjug** [comment]
@alexsoble you're still showing as changes requested, can you approve if everything looks good to you?

**Comment by amymok** [comment]
It looks like Arrangement Type and OPRE Division are showing indices
Screen Shot 2021-09-24 at 10 14 09 AM

I am also getting some errors in migration when I bring up the system.
```
➜  OPRE-OPS git:(cj/93/view-cans) ✗ docker-compose up   
Starting opre-ops_db_1 ... done
Starting opre-ops_migration_1 ... done
Starting opre-ops_web_1       ... done
Attaching to opre-ops_db_1, opre-ops_migration_1, opre-ops_web_1
migration_1  | checking for database on db:5432
migration_1  | Connection to db (172.18.0.2) 5432 port [tcp/postgresql] succeeded!
migration_1  | found database on db:5432!
db_1         | 
db_1         | PostgreSQL Database directory appears to contain a database; Skipping initialization
db_1         | 
db_1         | 2021-09-24 17:16:14.280 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1         | 2021-09-24 17:16:14.280 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1         | 2021-09-24 17:16:14.285 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1         | 2021-09-24 17:16:14.303 UTC [26] LOG:  database system was interrupted; last known up at 2021-09-24 17:12:34 UTC
db_1         | 2021-09-24 17:16:14.596 UTC [26] LOG:  database system was not properly shut down; automatic recovery in progress
db_1         | 2021-09-24 17:16:14.599 UTC [26] LOG:  redo starts at 0/17891C0
db_1         | 2021-09-24 17:16:14.599 UTC [26] LOG:  invalid record length at 0/17892A0: wanted 24, got 0
db_1         | 2021-09-24 17:16:14.599 UTC [26] LOG:  redo done at 0/1789268
db_1         | 2021-09-24 17:16:14.614 UTC [1] LOG:  database system is ready to accept connections
db_1         | 2021-09-24 17:16:14.796 UTC [33] LOG:  incomplete startup packet
migration_1  | Operations to perform:
migration_1  |   Apply all migrations: admin, auth, contenttypes, ops_site, sessions
migration_1  | Running migrations:
db_1         | 2021-09-24 17:16:15.446 UTC [35] ERROR:  relation "ops_site_caninfo" already exists
db_1         | 2021-09-24 17:16:15.446 UTC [35] STATEMENT:  CREATE TABLE "ops_site_caninfo" ("id" bigserial NOT NULL PRIMARY KEY, "number" varchar(30) NOT NULL, "description" varchar(100) NOT NULL, "purpose" text NOT NULL, "nickname" varchar(30) NOT NULL, "arrangement_type" varchar(30) NOT NULL, "authorizer_id" bigint NOT NULL)
migration_1  |   Applying ops_site.0007_auto_20210924_1700...Traceback (most recent call last):
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 82, in _execute
migration_1  |     return self.cursor.execute(sql)
migration_1  | psycopg2.errors.DuplicateTable: relation "ops_site_caninfo" already exists
migration_1  | 
migration_1  | 
migration_1  | The above exception was the direct cause of the following exception:
migration_1  | 
migration_1  | Traceback (most recent call last):
migration_1  |   File "/opre_project/manage.py", line 23, in <module>
migration_1  |     main()
migration_1  |   File "/opre_project/manage.py", line 19, in main
migration_1  |     execute_from_command_line(sys.argv)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
migration_1  |     utility.execute()
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
migration_1  |     self.fetch_command(subcommand).run_from_argv(self.argv)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
migration_1  |     self.execute(*args, **cmd_options)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 398, in execute
migration_1  |     output = self.handle(*args, **options)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 89, in wrapped
migration_1  |     res = handle_func(*args, **kwargs)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 244, in handle
migration_1  |     post_migrate_state = executor.migrate(
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 117, in migrate
migration_1  |     state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
migration_1  |     state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
migration_1  |     state = migration.apply(state, schema_editor)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/migrations/migration.py", line 126, in apply
migration_1  |     operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/migrations/operations/models.py", line 92, in database_forwards
migration_1  |     schema_editor.create_model(model)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 345, in create_model
migration_1  |     self.execute(sql, params or None)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 145, in execute
migration_1  |     cursor.execute(sql, params)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 98, in execute
migration_1  |     return super().execute(sql, params)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
migration_1  |     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
migration_1  |     return executor(sql, params, many, context)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
migration_1  |     return self.cursor.execute(sql, params)
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
migration_1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
migration_1  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 82, in _execute
migration_1  |     return self.cursor.execute(sql)
migration_1  | django.db.utils.ProgrammingError: relation "ops_site_caninfo" already exists
migration_1  | 
opre-ops_migration_1 exited with code 1
web_1        | Watching for file changes with StatReloader
web_1        | Performing system checks...
web_1        | 
web_1        | System check identified no issues (0 silenced).
web_1        | 
web_1        | You have 1 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): ops_site.
web_1        | Run 'python manage.py migrate' to apply them.
web_1        | September 24, 2021 - 17:16:16
web_1        | Django version 3.2.6, using settings 'opre_ops.settings.local'
web_1        | Starting development server at http://0.0.0.0:8080/
web_1        | Quit the server with CONTROL-C.
```

**Comment by amymok** [comment]
Okay sorry, it looks like I didn't clean the docker container.  It is good after I do `docker-compose down` then re-build/up.

LGTM!

**Review by alexsoble — APPROVED** [review]
LGTM! 🚀  I will 🚢 since Amy and Nina have both approved!

**Pull request merged**
